### ISSUE TITLE
creates temporary files badly

### ISSUE THREAD

**Issue by jmberg** [AFLplusplus/AFLplusplus#366]
AFL++ creates temporary files like this:

```
  modified_file =
      alloc_printf("%s/.afl-%u-%u.s", tmp_dir, getpid(), (u32)time(NULL));
```

That's bad, if the machine is fast enough to recycle the PID within a second, things clash. Happened here, I believe.

**Comment by vanhauser-thc** [comment]
Hi, the code you are quoting is from "afl-as.c" which is very likely not where your issue is.
But I get your point. workaround is easy, should land in the dev branch today.

**Comment by jmberg** [comment]
Or maybe not, I see it uses `O_WRONLY | O_EXCL | O_CREAT`.

I've replaced that code above though, will see if the issues persist.

**Comment by jmberg** [comment]
Well, I am still using afl-gcc (yeah, will fix that soon), and am seeing random and unexplained linker errors, so I thought maybe the files were clashing here. But with O_EXCL it should run into a PFATAL() at that point, so now I'm not so sure...

Need to watch this a bit more I guess.

**Comment by vanhauser-thc** [comment]
I am performing an online training at this moment (break time), thinking about your problem, the pid is only used in filenames for compiling. so it is really just an issue there. highly unlike to happen, but can.

**Comment by jmberg** [comment]
I thought we'd have a conversation via a github issue anyway :-)

I agree it's not likely, but I have some very fast build machines at hand (96 CPU cores or so), so they do run through these things very quickly. And a significant percentage of our builds see highly random linker errors, so I started looking around.

This _seemed_ like a candidate until I saw the O_EXCL, and since we're not seeing the PFATAL() it's quite possibly something else ... but I've fixed it anyway and will continue to check.

**Comment by vanhauser-thc** [comment]
checked in, please test

**Comment by vanhauser-thc** [comment]
a FATAL would still be a problem for building.

**Comment by jmberg** [comment]
yes, but it would explain what was going on, and not result in broken binaries...

patch looks good, but I can't really test it quickly since the problem is intermittent.

**Comment by jmberg** [comment]
Thanks!

**Comment by jmberg** [comment]
FWIW, this wasn't my issue.

And I'm not actually sure what I said made sense - the pid cannot be recycled while we're still running, and the only use of the file is while we're running ... Sorry for the noise.

**Comment by jmberg** [comment]
(of course, on a sufficiently fast machine this might still matter for AFL_KEEP_ASSEMBLY, but not for me)

**Comment by vanhauser-thc** [comment]
> And I'm not actually sure what I said made sense - the pid cannot be recycled while we're still running, and the only use of the file is while we're running ... Sorry for the noise.

well it could happen as afl-as generates the file, but the linker is linking it. so in theory it is possible that another afl-as run received the same pid and then generates the same file. think "make -j 4" and a highly process volatile system.

**Comment by jmberg** [comment]
Yeah, I thought so too, but that's not what happens.

What happens is this
1. gcc is invoked
1. gcc invokes cc1 on test.c -o /tmp/ccXXXXXX.s, that name is picked by gcc, cc1 finishes
1. gcc invokes as (afl-as) -o /tmp/ccXXXXXY.o /tmp/ccXXXXXX.s
1. afl-as creates its temp file, invokes the real as, that writes the result to the file gcc picked - /tmp/ccXXXXXY.o, and, most importantly, afl-as waits for it to finish. At this point, afl-as's temp file is no longer relevant or used again.
1. now collect2 and later ld come in, but they only work on the filename gcc picked

So, I got confused, like I said - sorry for the noise.

I still seems to have issues anyway, which sort of empirically proves this too.

That said, it *is* a very busy system, and it's not like "make -j4", more like "make -j96" :-)

**Comment by jmberg** [comment]
FWIW, I think I figured it out. What was going on is that part of the build system is not written in make, and different sub-projects are built in parallel (but outside make). Now several of the sub-projects require the same library, and if those sub-projects are built in parallel, the system will attempt to build the same library (in the same filesystem location and all) multiple times in parallel.

We had dealt with this for the non-fuzzing parts of the system properly by compiling the library up-front, but I forgot to do that for the fuzzing parts ...

So, completely an issue on my side, but I guess at least theoretically we made AFL_KEEP_ASSEMBLY slightly more robust... :)
